src/compiler.rs

```diff
@@ -0,0 +1,187 @@
+use std::path::Path;
+use std::sync::Arc;
+
+use await_tree::InstrumentAwait;
+use log::error;
+
+use typst::diag::SourceResult;
+use typst::layout::Position;
+use typst::model::Document;
+use typst::syntax::Span;
+
+use typst_ts_compiler::service::{
+    CompileActor, CompileClient as TsCompileClient, CompileExporter, Compiler, WorldExporter,
+};
+use typst_ts_compiler::service::{CompileDriver, CompileMiddleware};
+use typst_ts_compiler::vfs::notify::{FileChangeSet, MemoryEvent};
+use typst_ts_core::debug_loc::SourceSpanOffset;
+use typst_ts_core::error::prelude::ZResult;
+
+use typst_preview::{CompilationHandle, CompileStatus};
+use typst_preview::{CompileHost, EditorServer, MemoryFiles, MemoryFilesShort, SourceFileServer};
+use typst_preview::{DocToSrcJumpInfo, Location};
+
+pub type CompileService<H> = CompileActor<Reporter<CompileExporter<CompileDriver>, H>>;
+pub type CompileClient<H> = TsCompileClient<CompileService<H>>;
+
+pub struct CompileServer<H: CompilationHandle> {
+    inner: CompileService<H>,
+    client: TypstClient<H>,
+}
+
+pub struct Reporter<C, H> {
+    inner: C,
+    cb: H,
+}
+
+impl<C: Compiler, H: CompilationHandle> CompileMiddleware for Reporter<C, H> {
+    type Compiler = C;
+
+    fn inner(&self) -> &Self::Compiler {
+        &self.inner
+    }
+
+    fn inner_mut(&mut self) -> &mut Self::Compiler {
+        &mut self.inner
+    }
+
+    fn wrap_compile(
+        &mut self,
+        env: &mut typst_ts_compiler::service::CompileEnv,
+    ) -> SourceResult<Arc<Document>> {
+        self.cb.status(CompileStatus::Compiling);
+        match self.inner_mut().compile(env) {
+            Ok(doc) => {
+                self.cb.notify_compile(Ok(doc.clone()));
+                Ok(doc)
+            }
+            Err(err) => {
+                self.cb.notify_compile(Err(CompileStatus::CompileError));
+                Err(err)
+            }
+        }
+    }
+}
+
+impl<C: Compiler + WorldExporter, H> WorldExporter for Reporter<C, H> {
+    fn export(&mut self, output: Arc<typst::model::Document>) -> SourceResult<()> {
+        self.inner.export(output)
+    }
+}
+
+impl<H: CompilationHandle> CompileServer<H> {
+    pub fn new(
+        compiler_driver: CompileDriver,
+        cb: H,
+        // renderer_sender: broadcast::Sender<RenderActorRequest>,
+        // editor_conn_sender: mpsc::UnboundedSender<EditorActorRequest>,
+    ) -> Self {
+        // CompileExporter + DynamicLayoutCompiler + WatchDriver
+        let root = compiler_driver.world.root.clone();
+        // let r = renderer_sender.clone();
+        let driver = CompileExporter::new(compiler_driver);
+        let driver = Reporter { inner: driver, cb };
+        let inner = CompileActor::new(driver, root.as_ref().to_owned()).with_watch(true);
+
+        Self {
+            inner,
+            client: TypstClient {
+                inner: once_cell::sync::OnceCell::new(),
+            },
+        }
+    }
+
+    pub fn spawn(self) -> ZResult<TypstClient<H>> {
```

Comment:
what is ZResult?

Reply:
`ZResult` is defined in typst.ts/core/error. It is `Result<T, typst_ts_core::Error>`

Reply:
hmm.. The name is not good. anyway, it's not a big deal

Reply:
I don't like `type Result<T> = std::Result<T, ts::Error>`, neither `typst TsResult<T> = ...`, so I add one extra alphabet to reduce redundancy.

Reply:
I have inlined `ZResult` in typst_preview. it is heavily used in ts but not preview, so we can tolerate the redundancy.